core/src/main/scala/org/apache/spark/scheduler/TaskResultGetter.scala

```diff
@@ -84,6 +90,7 @@ private[spark] class TaskResultGetter(sparkEnv: SparkEnv, scheduler: TaskSchedul
               }
               val deserializedResult = serializer.get().deserialize[DirectTaskResult[_]](
                 serializedTaskResult.get.toByteBuffer)
+              deserializedResult.value(taskResultSerializer.get())
```

Comment:
nit: Would be nice to add a comment here saying "force deserialization of referenced value" or some such